kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/filter/NettyFilterDispatchExecutor.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal.filter;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import io.netty.channel.EventLoop;
+
+import io.kroxylicious.proxy.filter.FilterDispatchExecutor;
+
+class NettyFilterDispatchExecutor implements FilterDispatchExecutor {
+
+    private final EventLoop eventLoop;
+
+    NettyFilterDispatchExecutor(EventLoop eventLoop) {
+        Objects.requireNonNull(eventLoop, "eventLoop cannot be null");
+        this.eventLoop = eventLoop;
+    }
+
+    @Override
+    public boolean isInFilterDispatchThread() {
+        return eventLoop.inEventLoop();
+    }
+
+    EventLoop getEventLoop() {
+        return eventLoop;
+    }
+
+    @Override
+    public <T> CompletionStage<T> completeOnFilterDispatchThread(CompletionStage<T> completionStage) {
+        CompletableFuture<T> future = new CompletableFuture<>();
+        completionStage.whenComplete((value, throwable) -> {
+            if (isInFilterDispatchThread()) {
+                forward(value, throwable, future);
+            }
+            else {
+                execute(() -> forward(value, throwable, future));
+            }
+        });
+        return future;
+    }
+
+    private static <T> void forward(T value, Throwable throwable, CompletableFuture<T> future) {
+        if (throwable != null) {
+            future.completeExceptionally(throwable);
+        }
+        else {
+            future.complete(value);
+        }
+    }
+
+    @Override
+    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+        return eventLoop.schedule(command, delay, unit);
+    }
+
+    @Override
+    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
+        return eventLoop.schedule(callable, delay, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit) {
+        return eventLoop.scheduleAtFixedRate(command, initialDelay, period, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay, TimeUnit unit) {
+        return eventLoop.scheduleWithFixedDelay(command, initialDelay, delay, unit);
+    }
+
+    @Override
+    public void shutdown() {
+        throw new UnsupportedOperationException("filter dispatch thread cannot be shutdown by clients");
```

Comment:
Won't this break try-with-resources as close calls shutdown?  I think I'd make shutdown a no-op.